### ISSUE TITLE
Average- and MaxPooling behave strangely

### ISSUE THREAD

**Issue by HughRunyan** [facebookresearch/SparseConvNet#140]
I'm trying to build a squeeze/excite-style block. Unpooling is undesirable because of all the blank space, but it runs without complaint. However, Deconvolution after Average- or MaxPooling fails. It returns tensor.features of shape [0, N_features] where the 0 was previously the number of input points. AveragePooling outputs features that don't look obviously crazy and spatial_size works correctly for all of the above, but tensor.get_spatial_locations() is empty after pooling. I believe that is the reason the Deconvolution layer fails but I'm not positive. Strangely tensor.metadata.getSpatialLocations(tensor.spatial_size) is empty, but tensor.metadata.getSpatialLocations(tensor.spatial_size before pool operation) seems to give the desired answer. I can't find any obvious crossed wires in the python wrappers so it seems likely the problem is in the metadata or pooling CUDA files. I'll update if I find it.

**Comment by btgraham** [comment]
Can you provide a minimal example? This simple example works as expected for me.
```
import sparseconvnet as scn
scn.Sequential(
    scn.InputLayer(2,32,0),
    scn.AveragePooling(2,2,2),
    scn.Deconvolution(2,1,1,2,2,False)
)((
    torch.LongTensor([[1,2,0],[2,3,0],[1,2,1]]),
    torch.randn(2,1)
))
```

**Comment by HughRunyan** [comment]
'''

class SE(nn.Module):
    def __init__(self):
        nn.Module.__init__(self)

        self.inp = scn.InputLayer(3, 1024, mode=4)
        self.c1 = scn.SubmanifoldConvolution(3, 3, 32, 3, False)
        self.bn = scn.BatchNormLeakyReLU(32)
        self.ap = scn.MaxPooling(3, 8, 8)
        self.c2 = scn.SubmanifoldConvolution(3, 32, 32, 3, False)
        self.dc = scn.Deconvolution(3, 32, 32, 3, 3, False))

    def forward(self,x):
        
        x = self.inp(x)
        x = self.c1(x)
        x2 = self.bn(x)
        x2 = self.ap(x2)
        x2 = self.c2(x2)
        x2 = self.bn(x2)
        x2 = self.dc(x2)

        return [x, x2]
'''

produces the undesired effect in x2 on my system. Same is true if it's built as a sequential model with the same structure as the UNet blocks in the networkArchitectures.py file.

If I alter the above code slightly to save the output of every layer in a list called predictions instead of just the two, this is the output.

'''
for i in np.arange(len(predictions)):
    print(predictions[i].metadata.getSpatialLocations(predictions[i].spatial_size).shape)

torch.Size([1174457, 4])
torch.Size([1174457, 4])
torch.Size([0, 4])
torch.Size([0, 4])
torch.Size([0, 4])
torch.Size([0, 4])

for i in np.arange(len(predictions)):
    print(predictions[i].get_spatial_locations())

 tensor([[109, 270, 371,   0],
        [142, 320, 418,   0],
        [126, 316, 419,   0],
        ...,
        [640,  94,  54,   7],
        [457,  79,  82,   7],
        [637, 132,  66,   7]])
tensor([[109, 270, 371,   0],
        [142, 320, 418,   0],
        [126, 316, 419,   0],
        ...,
        [640,  94,  54,   7],
        [457,  79,  82,   7],
        [637, 132,  66,   7]])
tensor([], size=(0, 4), dtype=torch.int64)
tensor([], size=(0, 4), dtype=torch.int64)
tensor([], size=(0, 4), dtype=torch.int64)
tensor([], size=(0, 4), dtype=torch.int64)

# note the predictions[0].spatial_size here

for i in np.arange(len(predictions)):
    print(predictions[i].metadata.getSpatialLocations(predictions[0].spatial_size).shape)

torch.Size([1174457, 4])
torch.Size([1174457, 4])
torch.Size([1174457, 4])
torch.Size([1174457, 4])
torch.Size([1174457, 4])
torch.Size([1174457, 4])

```

**Comment by HughRunyan** [comment]
For reference, the 6 tensors are outputs of layers c1 -> dc respectively. I didn't save the output of the input layer, which is why there's only 6

**Comment by HughRunyan** [comment]
While I have you...I'm soon going to try building a version of SCN that ingests aligned higher-res depth maps of a 3D scene as well as the reference point cloud I want segmented, stacks the feature maps, and outputs a prediction for the point cloud points but not all the extra depth map points. I have high-res imagery and camera pose info from the photogrammetric model-building process I want to use. Any tips on how to crack open the SCN tensor object and do feature map concatenation without messing up the feature map grid/point cloud points correspondence (or anything else) would be appreciated.

**Comment by btgraham** [comment]
A deconvolution has to undo a previous convolution, as in a UNet.
```
        self.ap = scn.MaxPooling(3, 8, 8)
        self.dc = scn.Deconvolution(3, 32, 32, 3, 3, False))
```
You are using size=8, stride=8 max-pooling with a size=3, stride=3 deconvolution. This is not supported - the size and stride have to match so input sparsity pattern can match the output one.

**Comment by btgraham** [comment]
> While I have you...I'm soon going to try building a version of SCN that ingests aligned higher-res depth maps of a 3D scene as well as the reference point cloud I want segmented, stacks the feature maps, and outputs a prediction for the point cloud points but not all the extra depth map points. I have high-res imagery and camera pose info from the photogrammetric model-building process I want to use. Any tips on how to crack open the SCN tensor object and do feature map concatenation without messing up the feature map grid/point cloud points correspondence (or anything else) would be appreciated.

So you are doing structure from motion? Projecting all the depths maps into the same input space is likely to produce weird shapes, like the same wall appearing twice but separated by 1cm. This pushes up the number of active locations hurting efficiency.  You might be better off using something like tsdf-fusion to merge them before putting them into SparseConvNet.

If you have a SparseConvNetTensor, and you want to extract a subset of the points, either use .get_spatial_locations(), or if speed is critical there could be a layer to extract features corresponding to a  (n,3) tensor of coordinates.

**Comment by HughRunyan** [comment]
re the deconvolution: in my original program I had three consecutive 2/2 deconvolutions to return it to the original size and checked with .spatial_size() to confirm it returned to same, but took them out while hunting for what was going wrong. Does that arrangement not work with the way the library is constructed? I used 8 because that's what the paper uses, and as the point is to get a broader contextual view 2x2 seemed too light a touch. The same effect could be achieved by some 3x3 convs or whatever, probably plenty of other ways to do it, I just wasn't sure why this one wasn't working. 

Re structure from motion: yes the models are built with SfM. But we take 18mm and 55mm photos in parallel, use both to build the "sparse cloud", if you're familiar with terminology from that field, but only the 18s to do the final dense point cloud because the 55s are unreliable, often produce holes or large-scale geometric drift. If we took more pictures with more overlap that could probably be rectified in at least most cases, but our field procedure is pretty rough as it is and we can't go back in time and fix the existing ones. Hopefully at some point in the future this will not be how it is, but we already have a large database of models built with the 18s. Because of that sparse cloud step though, we do have alignment/depth data for all the 55s. 

I will investigate your tsdf tip it looks promising. To avoid the overlapping surface ghosting type problem you referenced, I was planning on passing the network, say, 4 depth maps one at a time from the 55s at a higher resolution than my standard unet, do a couple convs accumulate the feature maps, and then concatenate them to the feature maps from the first conv layer of the standard unet. Start at I dunno 1 mm resolution, use 4096x4096x4096, step down to 1024x1024x1024x8 or whatever, stack four of them 1024x1024x1024x32, output of the first unet conv layer for the normal point cloud is 1024x1024x1024x32, whack them together 1024x1024x1024x64. Hi-res data gets transferred. Could do it any number of ways, that one seems pretty straightforward. But I don't see a way to do that feature-grid-level concatenation in the library, just the .features list from the points, and doing the above by matching points across the clouds and transferring features seems a lot slower/more cumbersome. So I have been operating under the assumption I would have to alter build a cuda function to do it and thought you might have some tips as to where to start. 

Thanks for your time and help by the way, much appreciated, apologies for the hassle and wordy queries.

**Comment by HughRunyan** [comment]
I'd have to drop features from the depth maps that correspond to non-active sites in the normal point cloud.

**Comment by HughRunyan** [comment]
Oh I remember why that deconv in the example code I sent you is 3/3 instead of 2/2, I was messing with it to see if anything worked. I previously tried to make the unet steeper by changing your 2/2 conv/deconvs to 3/3, 4/4, etc and had trouble getting it to work with odd numbers (3/3, 5/5), which is weird but probably my fault, some kind of matrix dimension problem somewhere, or odd numbers/padding or something, it wasn't critical so I never resolved it

**Comment by HughRunyan** [comment]
It seemed worth trying 3/3 even though that had previously been trouble...

**Comment by HughRunyan** [comment]
In your unet, I see convolutions and deconvs are indeed paired except having a sequential between them containing all the intervening steps. To do a squeeze-type operation, would I have to mimic that format, like, conv, sequential(conv, deconv), deconv? I thought deconvs were just heavily-padded convolutions like https://github.com/vdumoulin/conv_arithmetic (they use "transposed convolution"), I guess you're passing some information between the conv and deconv layers? Maybe because I think I remember reading somewhere there is no padding in your architecture?

**Comment by HughRunyan** [comment]
Nevermind you mention padding in the paper, seems like that's not it

**Comment by btgraham** [comment]
Lots of questions ..
Contact me at btgraham@gmail.com if you want to VC ..
